### PR TITLE
Linting Fixups

### PR DESCRIPTION
**Pull request by MightyAx**
All our UI linting fixups in one place.
I only changed components I understood, so dependencies like the clipboard were left alone.

~~The quicker we review this the better, because this would be a nightmare of a merge conflict.~~
**This one will need some testing because of deprecated RxJS functionality.**

Deployed to [Dev](https://dev.contribute.data.humancellatlas.org)

### CODE REVIEW

**Review by aaclan-ebi — APPROVED** [review]
This is soooo noice!

File: client/src/app/project-registration/contact-field-group/contact-field-group.component.spec.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ describe('ContactFieldGroupComponent', () => {
   beforeEach(() => {
     userSpy = jasmine.createSpyObj(['subscribe']) as jasmine.SpyObj<Subject<User>>;
     firstUserSpy = jasmine.createSpyObj(['first']) as jasmine.SpyObj<Subject<User>>;
-    firstUserSpy.first.and.returnValue(userSpy);
+    // firstUserSpy.first.and.returnValue(userSpy); I don't know how to fix this!
```
Comment (MightyAx) [inline client/src/app/project-registration/contact-field-group/contact-field-group.component.spec.ts:28]: > This is soooo noice!

It was, it was also terribly broken because of a change in RxJS.
I've updated all our usages but we'll have to be careful with this one, lots of testing.

Also re: unit tests, I don't know what to do about this one.
The format is completely foreign to me.

**Comment by MightyAx** [comment]
For Investigation:
- [ ] Might just be my machine but there seems to be a delay what querying My Projects, All Projects, All Submissions.

File: client/src/app/my-projects/my-projects.component.ts [hunk 2]
```diff
@@ -55,12 +54,8 @@ export class MyProjectsComponent implements OnInit, OnDestroy, AfterViewInit {
   ngOnInit() {
     this.isLoggedIn$ = this.aai.isUserLoggedIn();
     this.account$ = this.isLoggedIn$.pipe(
-      concatMap(loggedIn => {
-        if (loggedIn) {
-          return this.ingestService.getUserAccount();
-        }
-        return of(undefined);
-      })
+      filter(loggedIn => loggedIn.valueOf()),
```
Comment (rolando-ebi) [inline client/src/app/my-projects/my-projects.component.ts:57]: For brevity, could just do `filter(loggedIn => loggedIn)` or `filter(identity)` (would need to import `identity` from rxjs). 

But for readability sake, I'd just do `filter(loggedIn => loggedIn === true)`

File: client/src/app/aai/aai.service.ts [hunk 3]
```diff
@@ -31,18 +33,18 @@ export class AaiService {
   }
 
   getUser(): Observable<User> {
-    return from(this.manager.getUser()).map(user => {
+    return from(this.manager.getUser()).pipe(map(user => {
```
Comment (rolando-ebi) [inline client/src/app/aai/aai.service.ts:36]: I think by convention, operators go on their own lines in a pipe e.g

```
const $stream = of([1,2,3,4,5]).pipe(
   map(num => num * 2),
   ...
   ...
);
```

File: client/src/app/shared/services/ontology.service.ts [hunk 4]
```diff
@@ -49,18 +52,20 @@ export class OntologyService {
...
-      .map(olsResponse => olsResponse.docs[0].iri)
-    ).map(iriArray => {
-      searchParams[this.OLS_RELATION[ontologyRelation]] = iriArray;
-      return searchParams;
-    });
+    ).pipe(
+      map(responseArray => responseArray
+        .map(ols => ols.response)
+        .filter(olsResponse => olsResponse.numFound === 1)
+        .map(olsResponse => olsResponse.docs[0].iri)),
```
Comment (rolando-ebi) [inline client/src/app/shared/services/ontology.service.ts:59]: Some operators here aren't piped e.g `ontologyClasses.map..` and `responseArray.map....`

Comment (rolando-ebi) [reply client/src/app/shared/services/ontology.service.ts:59]: This occurs a few times (could do a Find-> .map for all occurrences)

**Review by rolando-ebi — APPROVED** [review]
Nice, LGTM

If we want all rxjs operators used to exist within `pipe()`'s, there's a TSLint plugin that can ensure this: https://github.com/ReactiveX/rxjs-tslint

File: client/src/app/shared/services/ontology.service.ts (hunk 4 above)
Comment (MightyAx) [reply client/src/app/shared/services/ontology.service.ts:59]: those are maps from arrays rather than Observables

File: client/src/app/my-projects/my-projects.component.ts (hunk 2 above)
Comment (MightyAx) [reply client/src/app/my-projects/my-projects.component.ts:57]: > `filter(loggedIn => loggedIn)`

Wasn't allowed because of boolean != Boolean

I'll replace with `filter(loggedIn => loggedIn === true)`

File: client/src/app/shared/services/ontology.service.ts (hunk 4 above)
Comment (MightyAx) [reply client/src/app/shared/services/ontology.service.ts:59]: In which case is there a different convention to make that obvious?

Comment (rolando-ebi) [reply client/src/app/shared/services/ontology.service.ts:59]: Ahh I see, that's fine. Could also do `of(responseArray).pipe(....)` to get back into observables land. Could also look at flatMap()-ing the nested map()s

File: client/src/app/template-questionnaire/template-generator.service.ts [hunk 5]
```diff
@@ -35,28 +36,25 @@ export class TemplateGeneratorService {
...
+      filter(data => data._links.download.href.length > 0),
+      map(data => data._links.download.href),
+      distinctUntilChanged(),
+      switchMap(href => this._requestToDownload(href)),
+      timeoutWith(this.TIMEOUT, throwError(new Error('Retrieval of template spreadsheet has timed out.'))),
+      filter(template => template.complete),
+      take(1),
+      map(template => template.data)
+    );
   }
```
Comment (aaclan-ebi) [inline client/src/app/template-questionnaire/template-generator.service.ts:50]: wow, you are now the observables guru

File: client/tsconfig.json [hunk 6]
```diff
@@ -19,7 +19,6 @@
       "dom"
     ],
     "module": "esnext",
-    "baseUrl": "./",
-    "resolveJsonModule": true
```
Comment (aaclan-ebi) [inline client/tsconfig.json:23]: is this not needed to import the json files?

**Review by aaclan-ebi — APPROVED** [review]
Wow this is huge PR. Is this already deployed in dev? we should do manual tests and merge soon.

File: client/tsconfig.json (hunk 6 above)
Comment (MightyAx) [reply client/tsconfig.json:23]: it's a repeat of line four, the linter spotted it.

File: client/src/app/metadata-schema-form/metadata-field-types/ontology-base/ontology-base.component.ts [hunk 7]
```diff
@@ -26,10 +26,12 @@ export class OntologyBaseComponent extends BaseInputComponent implements OnInit
     super.ngOnInit();
     const value: Ontology = this.metadataFormService.cleanFormData(this.control.value);
     this.searchControl = this.createSearchControl(value);
+    // noinspection JSDeprecatedSymbols
```
[the remark this answers is not in the document]
Comment (MightyAx) [reply client/src/app/metadata-schema-form/metadata-field-types/ontology-base/ontology-base.component.ts:29]: It's a false positive deprecation warning, I'll add that the comments so the context isn't lost in future:
[ReactiveX/rxjs#4772](https://github.com/ReactiveX/rxjs/issues/4772)

File: client/src/app/template-questionnaire/template-generator.service.ts [hunk 8]
```diff
@@ -35,28 +36,25 @@ export class TemplateGeneratorService {
...
-      .flatMap(data =>
-        Observable.interval(this.POLLING_INTERVAL)
-          .flatMap(() => this._requestToDownload(data._links.download.href))
-          .filter(template => template.complete)
-          .take(1)
-          .map(template => template.data)
-          .timeoutWith(this.TIMEOUT,
-            throwError(new Error('Retrieval of template spreadsheet has timed out.'))
-          ));
+    return timer(0, this.POLLING_INTERVAL).pipe(
```
Comment (rolando-ebi) [inline client/src/app/template-questionnaire/template-generator.service.ts:39]: @MightyAx Quick question - it looks like the behaviour here is different to before, it attempts to generate a new spreadsheet every `POLLING_INTERVAL`, whereas the old code attempted to download the spreadsheet every `POLLING_INTERVAL`. Is this change intentional?

Comment (MightyAx) [reply client/src/app/template-questionnaire/template-generator.service.ts:39]: Yes good catch, I hadn't realised that the way that `_requestToGenerate` and `_requestToDownload` were linked once call to `_requestToGenerate` will require polling to `_requestToDownload` until the file is generated.

This was on my list to test though 😅

File: client/src/app/shared/guards/user-is-logged-in.guard.ts [hunk 9]
```diff
@@ -12,7 +11,7 @@ export class UserIsLoggedInGuard implements CanActivate {
   constructor(private aai: AaiService, private router: Router) {
   }
 
-  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean | UrlTree> {
-    return this.aai.isUserLoggedIn().map(loggedIn => loggedIn ? loggedIn : this.router.parseUrl(`/login?redirect=${encodeURIComponent(state.url)}`));
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): boolean | UrlTree {
+    return this.aai.userLoggedIn() ? true : this.router.parseUrl(`/login?redirect=${encodeURIComponent(state.url)}`);
```
Comment (MightyAx) [inline client/src/app/shared/guards/user-is-logged-in.guard.ts:15]: Is there a better way to write this?
Would this work?
```
return this.aai.userLoggedIn() || this.router.parseUrl(`/login?redirect=${encodeURIComponent(state.url)}`);
```

File: client/src/app/shared/services/ingest.service.ts [hunk 10]
```diff
@@ -161,43 +169,32 @@ export class IngestService {
     return this.http.get(url);
   }
 
+  public getAs<T>(url): Observable<T> {
+    return this.http.get(url).pipe(map(data => data as T));
```
Comment (MightyAx) [inline client/src/app/shared/services/ingest.service.ts:173]: Could change this to:
`return this.http.get<T>(url);`

File: client/src/app/project-registration/contact-field-group/contact-field-group.component.ts [hunk 11]
```diff
@@ -54,15 +53,16 @@ export class ContactFieldGroupComponent implements OnInit {
     // default
     correspondingCtrl.setValue(true);
 
-    this.aai.getUserSubject().pipe(first()).subscribe(user => {
-      this.userInfo = user ? user.profile : null;
+    const user = this.aai.getUser().value;
```
Comment (rolando-ebi) [inline client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: @MightyAx 
Is this a safe change to make? If the `this.aai.getUser()` subjects has no values at the time of initialisation then I assume `const user` will be null/undefined and then below code will never be called, or throw an error.

Comment (MightyAx) [reply client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: You can't get to the contact-field-group without logging in.

Comment (MightyAx) [reply client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: The call to `AaiService.loggedIn(user)` checks If the user is null

Comment (MightyAx) [reply client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: ah, do you mean the `.value` property?
BehaviourSubjects have an initial value supplied in the constructor
that initial value or whatever the latest value provided by the `.next` call is provided when you call `.value` or .`GetValue()`
```
private cachedUser$: BehaviorSubject<User> = new BehaviorSubject<User>(null);
this.cachedUser$.value //initially null
this.cachedUser$.next(<User>Bob) //will trigger call to any subscribers
this.cachedUser$.value // now Bob
this.cachedUser$.getValue() // Also Bob
```

Comment (rolando-ebi) [reply client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: The below code doesn't run in that case. Is that the behaviour we want, or do we want the below code to run at some point, for example when the user subject does have a value?

Comment (MightyAx) [reply client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: Yeah, but if the user hasn't logged in we can't get their name and email from their user account, so it's correct that we don't run the code in the case that they haven't logged in.

Comment (rolando-ebi) [reply client/src/app/project-registration/contact-field-group/contact-field-group.component.ts:56]: That makes sense, but can't we just subscribe to the user subject so that we run the below code when a non-null user becomes available?


e.g

```
this.aai.getUser().pipe(
   filter(maybeUser => mayberUser !== null),
   take(1)
).subscribe(user => {
  // below code
})
```

The prior code is similar but

File: client/src/app/project-registration/project-id/project-id.component.ts [hunk 12]
```diff
@@ -35,7 +40,7 @@ export class ProjectIdComponent implements OnInit {
 
   technology: string;
   otherTechnology: string;
-  contributor: string;
+  contributor_name: string;
```
Comment (rolando-ebi) [inline client/src/app/project-registration/project-id/project-id.component.ts:43]: camelCase?

Comment (MightyAx) [reply client/src/app/project-registration/project-id/project-id.component.ts:43]: that was part of #54 that was already merged, brought in so I could resolve merge conflicts.

**Comment by aaclan-ebi** [comment]
The submission details page is not loading.

**Comment by MightyAx** [comment]
> The submission details page is not loading.

Can you post a link @aaclan-ebi? This submission is working for me.
https://dev.contribute.data.humancellatlas.org/submissions/detail?uuid=101b61ed-a7c6-4c76-859b-ad8fcdcdafe4&project=ae6029a8-7446-45fa-ace3-794409cc3656

**Comment by aaclan-ebi** [comment]
> > The submission details page is not loading.
> 
> Can you post a link @aaclan-ebi? This submission is working for me.
> https://dev.contribute.data.humancellatlas.org/submissions/detail?uuid=101b61ed-a7c6-4c76-859b-ad8fcdcdafe4&project=ae6029a8-7446-45fa-ace3-794409cc3656
Ah that one's loading when i cleared my cache.

However, I still found a submission which is not loading:
https://dev.contribute.data.humancellatlas.org/submissions/detail?uuid=92cb09ba-d7a4-4c1f-a6cb-56e1c0464357 

I also noticed that it would take me to login page first for sometime before it redirects me to the submission details page.

File: client/src/app/shared/guards/user-is-logged-in.guard.ts (hunk 9 above)
Comment (rolando-ebi) [reply client/src/app/shared/guards/user-is-logged-in.guard.ts:15]: I think that would work:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR

```
o6 = false || 'Cat'      // f || t returns "Cat"
```
 And it looks like typescript correctly identifies the type of the expression to be `boolean | UrlTree`

**Comment by MightyAx** [comment]
Thanks @aaclan-ebi:
I see these errors in the Javascript console that I imagine are the culprits, I'm investigating
```
[Error] ERROR – TypeError: undefined is not an object (evaluating 'e._links')
TypeError: undefined is not an object (evaluating 'e._links')(anonymous function)

[Error] ERROR – TypeError: null is not an object (evaluating 'this.submissionEnvelopeId.length')
TypeError: null is not an object (evaluating 'this.submissionEnvelopeId.length')
	dr (main.4cfd5975e411d38be33b.js:1:3015587)
	handleError (main.4cfd5975e411d38be33b.js:1:3015758)
	next (main.4cfd5975e411d38be33b.js:1:3107854)
```

**Comment by MightyAx** [comment]
> However, I still found a submission which is not loading:
> https://dev.contribute.data.humancellatlas.org/submissions/detail?uuid=92cb09ba-d7a4-4c1f-a6cb-56e1c0464357

Can confirm fixed on dev

**Comment by aaclan-ebi** [comment]
> > However, I still found a submission which is not loading:
> > https://dev.contribute.data.humancellatlas.org/submissions/detail?uuid=92cb09ba-d7a4-4c1f-a6cb-56e1c0464357
> 
> Can confirm fixed on dev

Yes, it's loading now! thanks for fixing it!

> I also noticed that it would take me to login page first for sometime before it redirects me to the submission details page.

Did you get a chance to replicate this?

**Comment by MightyAx** [comment]
Reproduction:
Screen Recording 2020-10-08 at 12 16 19

**Review by aaclan-ebi — APPROVED** [review]
Thanks for spending so much effort on cleaning up our UI code! 

I've reviewed the changes. It's a big change, I tried to navigate around in dev, it seems to work the same as before. If we find any issues, let's just fix it as we go. Basic submission flow seems to work fine. Let's merge this soon :)
